### PR TITLE
BUG: Display precision doesn't affect complex float numbers #25514

### LINKED ISSUE #25514: Display precision doesn't affect complex float numbers
#### Code Sample

```python
import numpy as np
import pandas as pd

pd.set_option("display.precision", 3)

A = np.random.rand(3,3)
df_A = pd.DataFrame(A)

B = np.random.rand(3,3)+1j*np.random.rand(3,3)
df_B = pd.DataFrame(B)

print(df_A)
print(df_B)

```

**Output:**

```
    0      1      2
0  0.665  0.824  0.953
1  0.341  0.078  0.408
2  0.945  0.411  0.078
                                            0  \
0    (0.2553045397887609+0.8464631779709604j)   
1  (0.5724405361700162+0.047300152111596105j)   
2     (0.5824766758024993+0.974755917539835j)   

                                           1  \
0  (0.3533797280121552+0.20515824034905894j)   
1    (0.928312255311234+0.3909443194373995j)   
2   (0.2700966337410404+0.9137483492310767j)   

                                           2  
0  (0.21748538025578568+0.7730350615682122j)  
1  (0.2435709287109903+0.46366472945887327j)  
2   (0.6531622218102532+0.5102462678094016j)
```

#### Problem description

Display precision (`pd.set_option("display.precision", 3)`) doesn't affect complex float numbers like shown in the example.

#### Expected Output

```
    0      1      2
0  0.665  0.824  0.953
1  0.341  0.078  0.408
2  0.945  0.411  0.078

               0               1               2
0   0.976+0.935j    0.739+0.851j    0.436+0.734j
1   0.998+0.867j    0.774+0.849j    0.553+0.749j
2   0.405+0.049j    0.965+0.912j    0.292+0.958j

```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.7.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 142 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.4
pytest: None
pip: 18.1
setuptools: 39.0.1
Cython: 0.29
numpy: 1.15.4
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 7.1.1
sphinx: None
patsy: None
dateutil: 2.7.5
pytz: 2018.9
blosc: None
bottleneck: 1.2.1
tables: None
numexpr: None
feather: None
matplotlib: 3.0.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by sakarpanta**
- [x] closes #25514
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -2498,12 +2498,12 @@ def memory_usage(self, index=True, deep=False):
         ...              for t in dtypes])
         >>> df = pd.DataFrame(data)
         >>> df.head()
-           int64  float64  complex128 object  bool
-        0      1      1.0      (1+0j)      1  True
-        1      1      1.0      (1+0j)      1  True
-        2      1      1.0      (1+0j)      1  True
-        3      1      1.0      (1+0j)      1  True
-        4      1      1.0      (1+0j)      1  True
```
Comment (gfyoung) [inline pandas/core/frame.py]: Is this considered an improvement though?  Look at all of those zero's...

File: pandas/tests/io/formats/test_format.py [hunk 2]
```diff
@@ -1417,6 +1417,18 @@ def test_to_string_float_index(self):
...
+        df = DataFrame({'x': [
+            (0.4467846931321966 + 0.0715185102060818j),
+            (0.2739442392974528 + 0.23515228785438969j),
+            (0.26974928742135185 + 0.3250604054898979j)]})
+        result = df.to_string()
+        expected = ('                  x\n0  0.44678+0.07152j\n'
+                    '1  0.27394+0.23515j\n'
+                    '2  0.26975+0.32506j')
+        assert result == expected
+
```
Comment (gfyoung) [inline pandas/tests/io/formats/test_format.py]: What did this test used to output (before your changes)

Comment (sakarpanta) [reply pandas/tests/io/formats/test_format.py]: Before my changes the display.precision that is set to 5 would have been ineffective. The result would be exactly the same as the input.

File: pandas/core/frame.py (hunk 1 above)
Comment (sakarpanta) [reply pandas/core/frame.py]: There is no `display.precision` set for this, and displaying six digits after the decimal point for a float is default behavior, apparently. I'm not sure why the float64 doesn't appear as `1.000000` rather that `1.0` when generated in that way. When I generate a random float using Numpy `random`, or `random.uniform` I get the default six digit values.

Comment (gfyoung) [reply pandas/core/frame.py]: @jreback : Thoughts on this?

Comment (jreback) [reply pandas/core/frame.py]: you can step thru the code and see why the float displays as is. I *believe* this is correct. This is a pretty complex interaction of options.

**Review by WillAyd — CHANGES_REQUESTED** [review]
Can you add a whatsnew for v0.25?

File: pandas/core/frame.py (hunk 1 above)
Comment (jreback) [reply pandas/core/frame.py]: did you have a chance to figure this out?

**Comment by simonjayhawkins** [comment]
does this also change the notebook display? if so can we have a test for to_html() with notebook =True and False.

File: pandas/tests/io/formats/test_format.py [hunk 3]
```diff
@@ -1417,6 +1417,18 @@ def test_to_string_float_index(self):
                     '5.0  4')
         assert result == expected
 
+    def test_to_string_complex_float_formatting(self):
+        fmt.set_option('display.precision', 5)
```
Comment (simonjayhawkins) [inline pandas/tests/io/formats/test_format.py]: do we need to use the context manger here?

File: pandas/core/frame.py (hunk 1 above)
Comment (sakarpanta) [reply pandas/core/frame.py]: A little, yeah. Basically floats' zeros are being removed by the `_trim_zeros` method, where for every digit starting from the right hand side are being removed until we get a decimal or a non-zero number. We would need to modify this function to accept complex numbers as well.

File: pandas/tests/io/formats/test_format.py (hunk 3 above)
Comment (jreback) [reply pandas/tests/io/formats/test_format.py]: yep

File: pandas/core/frame.py (hunk 1 above)
Comment (jreback) [reply pandas/core/frame.py]: I think it might be better to do this for complex as well.

Comment (sakarpanta) [reply pandas/core/frame.py]: Update my PR with the fixes

**Comment by jreback** [comment]
can you merge master and see if you can update

**Comment by sakarpanta** [comment]
> can you merge master and see if you can update

Done.

File: pandas/core/frame.py [hunk 4]
```diff
@@ -2497,12 +2497,12 @@ def memory_usage(self, index=True, deep=False):
...
+           int64  float64   complex128  object  bool
```
Comment (jreback) [inline pandas/core/frame.py]: is there an extra space here?

File: pandas/io/formats/format.py [hunk 5]
```diff
@@ -894,6 +894,8 @@ def format_array(values, formatter, float_format=None, na_rep='NaN',
         fmt_klass = ExtensionArrayFormatter
     elif is_float_dtype(values.dtype):
```
Comment (jreback) [inline pandas/io/formats/format.py]: you can just do this in an or:

``is_float_dtype(...) or is_complex_dtype(...)`` easier to read

File: pandas/io/formats/format.py [hunk 6]
```diff
@@ -1424,6 +1430,21 @@ def just(x):
...
+        num_arr = str_complex.split('+')
+        return (_trim_zeros([num_arr[0]], na_rep) +
+                ['+'] +
+                _trim_zeros([num_arr[1][:-1]], na_rep) +
+                ['j'])
+
+    return [''.join(separate_and_trim(x, na_rep)) for x in str_complexes]
+
+
 def _trim_zeros(str_floats, na_rep='NaN'):
```
Comment (jreback) [inline pandas/io/formats/format.py]: can you rename to _trim_zeros_float

File: pandas/io/formats/format.py [hunk 7]
```diff
@@ -1094,7 +1097,10 @@ def format_values_with(float_format):
                                            for val in values.ravel()[imask]])
 
             if self.fixed_width:
-                return _trim_zeros(values, na_rep)
+                if is_complex:
```
Comment (jreback) [inline pandas/io/formats/format.py]: you can do the type check right here instead of above

File: pandas/tests/io/formats/test_format.py (hunk 3 above)
Comment (jreback) [reply pandas/tests/io/formats/test_format.py]: can you use the context manager here

File: pandas/tests/io/formats/test_format.py [hunk 8]
```diff
@@ -1370,6 +1370,18 @@ def test_to_string_float_index(self):
...
+        df = DataFrame({'x': [
+            (0.4467846931321966 + 0.0715185102060818j),
+            (0.2739442392974528 + 0.23515228785438969j),
+            (0.26974928742135185 + 0.3250604054898979j)]})
+        result = df.to_string()
+        expected = ('                  x\n0  0.44678+0.07152j\n'
```
Comment (jreback) [inline pandas/tests/io/formats/test_format.py]: do these have an extra space?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/formats/format.py (hunk 6 above)
Comment (sakarpanta) [reply pandas/io/formats/format.py]: Sure.

File: pandas/io/formats/format.py (hunk 7 above)
Comment (sakarpanta) [reply pandas/io/formats/format.py]: The problem with that is at that point the `values` object has been reassigned as `            values = np.array(values, dtype='object')` and is no longer a complex type but an array with the complex numbers as strings.

File: pandas/tests/io/formats/test_format.py (hunk 8 above)
Comment (sakarpanta) [reply pandas/tests/io/formats/test_format.py]: Yes. Good catch!

File: pandas/core/frame.py (hunk 4 above)
Comment (sakarpanta) [reply pandas/core/frame.py]: Yes. Good catch! Thanks.

File: pandas/tests/io/formats/test_format.py [hunk 9]
```diff
@@ -1370,6 +1370,18 @@ def test_to_string_float_index(self):
                     '5.0  4')
         assert result == expected
 
+    def test_to_string_complex_float_formatting(self):
+        with pd.option_context('display.precision', 5):
```
Comment (jreback) [inline pandas/tests/io/formats/test_format.py]: can you add the gh issue number as a comment

**Review by jreback — CHANGES_REQUESTED** [review]
minor change. ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @sakarpanta

**Pull request merged**